src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicSession.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Security;
+using System.Text;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal
+{
+    internal sealed class MsQuicSession : IDisposable
+    {
+        private bool _disposed = false;
+        private IntPtr _nativeObjPtr;
+        private MsQuicApi _api;
+        private bool _opened;
+
+        internal MsQuicSession(string registration)
+        {
+            _api = new MsQuicApi();
+
+            _api.RegistrationOpen(Encoding.UTF8.GetBytes(registration));
+        }
+
+        public MsQuicConnection ConnectionOpen(IPEndPoint endpoint, SslClientAuthenticationOptions sslClientAuthenticationOptions, IPEndPoint localEndpoint)
+        {
+            if (!_opened)
+            {
+                _nativeObjPtr = _api.SessionOpen(sslClientAuthenticationOptions.ApplicationProtocols[0].Protocol.ToArray());
+                _opened = true;
+                SetPeerBiDirectionalStreamCount(100);
+                SetPeerUnidirectionalStreamCount(100);
+            }
+
+            uint status = _api._connectionOpenDelegate(
+                _nativeObjPtr,
+                MsQuicConnection.NativeCallbackHandler,
+                IntPtr.Zero,
+                out IntPtr connectionPtr);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            return new MsQuicConnection(endpoint, _api, connectionPtr, sslClientAuthenticationOptions);
+        }
+
+        public MsQuicListener ListenerOpen(IPEndPoint listenEndPoint, SslServerAuthenticationOptions sslServerAuthenticationOptions)
```

Comment:
This design seems to limit you to a single certificate per listener, right? MsQuic allows for a design where you start the listener with no initial certificate, but as connections come in, you can "go get" the sec config based on the connection's requested server name (SNI). This allows a single listener for a given ALPN to support multiplexing multiple different host names with different certificates. Do you want to support this scenario?

Reply:
> Do you want to support this scenario?

Yes.

Reply:
We may need to add a callback here eventually to allow people to provide the cert they specifically want. Like what we did with ServerCertificateSelectionCallback in SslStream. I'd rather not design that in this PR, but it is definitely a scenario we need to support. I'll leave a TODO for now.

Reply:
SslServerAuthenticationOptions already has this, the ServerCertificateSelectionCallback.